airflow/models.py

```diff
@@ -2010,6 +2129,105 @@ def get(cls, key, session, deserialize_json=False):
         return v
 
 
+class XCom(Base):
+    """
+    Base class for XCom objects.
+    """
+    __tablename__ = "xcom"
+
+    id = Column(Integer, primary_key=True)
+    key = Column(String)
+    value = Column(PickleType(pickler=dill))
+    timestamp = Column(DateTime, server_default=func.current_timestamp())
+    visible_on = Column(DateTime, nullable=False)
+
+    # source information
+    task = Column(String, nullable=False)
+    dag = Column(String, nullable=False)
+
+    def __repr__(self):
+        return '<XCom "{key}" ({task} @ {visible_on})>'.format(
+            key=self.key,
+            task=self.task,
+            visible_on=self.visible_on)
+
+    @classmethod
+    @provide_session
+    def set(
+            cls,
+            key,
+            value,
+            visible_on,
+            task,
+            dag,
+            session=None):
+        """
+        Store an XCom value.
+        """
+
+        if not isinstance(task, basestring) or not isinstance(dag, basestring):
+            raise TypeError('Expected string; received {}'.format(type(task)))
+
+        session.expunge_all()
+        session.add(XCom(
+            key=key,
+            value=value,
+            visible_on=visible_on,
+            task=task,
+            dag=dag))
+        session.commit()
+
+    @classmethod
+    @provide_session
+    def get(
+            cls,
+            visible_on,
+            key=None,
+            tasks=None,
+            dags=None,
+            include_prior_dates=False,
+            limit=None,
+            session=None):
+        """
+        Retrieve an XCom value, optionally meeting certain criteria
+        """
+        filters = []
+        if key:
+            filters.append(cls.key == key)
+        if tasks:
+            filters.append(cls.task.in_(as_tuple(tasks)))
+        if dags:
+            filters.append(cls.dag.in_(as_tuple(dags)))
+        if include_prior_dates:
+            filters.append(cls.visible_on <= visible_on)
+        else:
+            filters.append(cls.visible_on == visible_on)
+
+        query = (
+            session.query(cls)
+            .filter(and_(*filters))
+            .order_by(cls.visible_on.desc(), cls.timestamp.desc())
+            .limit(limit or 1))
+
```

Comment:
Ok. the DataFrame was more useful in the previous (more complicated) version, where there were more fields to query. Let's drop it. And while we're at it, maybe we drop the whole ``limit`` concept entirely? It was great for debugging but I think it just confuses things -- I'm not sure what the real world use case would be. If users really want the XCom objects instead of values (though I'm not sure they will), then maybe a ``return_xcom=True`` argument could be added.

So there are two return types:
```python
ti.xcom_pull(key=optional, task_id=t) # returns a value
ti.xcom_pull(key=optional, task_id=[t1, t2, t3]) # returns a list of 3 values
```